server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthProviderConfigurator.java

```diff
@@ -48,31 +52,56 @@ protected OIDCIdentityProviderDefinition overlay(OIDCIdentityProviderDefinition
-            final AbstractXOAuthIdentityProviderDefinition definition) {
-        String authUrlBase;
-        if (definition instanceof OIDCIdentityProviderDefinition) {
-            authUrlBase = overlay((OIDCIdentityProviderDefinition) definition).getAuthUrl().toString();
-        } else {
-            authUrlBase = definition.getAuthUrl().toString();
-        }
-        String queryAppendDelimiter = authUrlBase.contains("?") ? "&" : "?";
-        List<String> query = new ArrayList<>();
-        query.add("client_id=" + definition.getRelyingPartyId());
-        query.add("response_type=" + URLEncoder.encode(definition.getResponseType(), StandardCharsets.UTF_8));
-        query.add("redirect_uri=" + URLEncoder.encode(baseURL + "/login/callback/" + alias, StandardCharsets.UTF_8));
-        query.add("state=" + uaaRandomStringUtil.getSecureRandom(10));
-        if (definition.getScopes() != null && !definition.getScopes().isEmpty()) {
-            query.add("scope=" + URLEncoder.encode(String.join(" ", definition.getScopes()), StandardCharsets.UTF_8));
+    public String getIdpAuthenticationUrl(
+            final AbstractXOAuthIdentityProviderDefinition definition,
+            final String idpOriginKey,
+            final HttpServletRequest request) {
+        var idpUrlBase = getIdpUrlBase(definition);
+        var callbackUrl = getCallbackUrlForIdp(idpOriginKey, UaaUrlUtils.getBaseURL(request));
+        var responseType = URLEncoder.encode(definition.getResponseType(), StandardCharsets.UTF_8);
+        var relyingPartyId = definition.getRelyingPartyId();
+
+        var state = generateStateParam();
+        saveStateParamForIdpToRequestSession(state, idpOriginKey, request);
+
+        UriComponentsBuilder uriBuilder = UriComponentsBuilder
+                .fromUriString(idpUrlBase)
+                .queryParam("client_id", relyingPartyId)
+                .queryParam("response_type", responseType)
+                .queryParam("redirect_uri", callbackUrl)
+                .queryParam("state", state);
+
+        if(!CollectionUtils.isEmpty(definition.getScopes())) {
+            uriBuilder.queryParam("scope", URLEncoder.encode(String.join(" ", definition.getScopes()), StandardCharsets.UTF_8));
         }
+
         if (OIDCIdentityProviderDefinition.class.equals(definition.getParameterizedClass())) {
-            final RandomValueStringGenerator nonceGenerator = new RandomValueStringGenerator(12);
-            query.add("nonce=" + nonceGenerator.generate());
+            var nonceGenerator = new RandomValueStringGenerator(12);
+            uriBuilder.queryParam("nonce", nonceGenerator.generate());
+        }
+
+        return uriBuilder.build().toUriString();
+    }
+
+    private void saveStateParamForIdpToRequestSession(String state, String idpOriginKey, HttpServletRequest request) {
+        if (request != null) {
+            request.getSession().setAttribute("xoauth-state-" + idpOriginKey, state);
+        }
```

Comment:
Is there any value to this null check? Even if it is null, this operation doesn't make any sense if we can't access the session.

Reply:
Good point